File: _scripts/audit-div-span-attributes.ts
// _scripts/audit-div-span-attributes.ts

// --- IMPORTS ---
import { walk } from "stdlib/fs";
import { join, dirname, fromFileUrl, relative } from "stdlib/path";

// --- CONFIGURATION ---
const scriptDir = dirname(fromFileUrl(import.meta.url));
const reportsDir = join(scriptDir, "../reports"); 
const luaFilter = join(scriptDir, "audit-div-span-attributes.lua");

// --- CATEGORIES ---
type Category = "MARK" | "ALT_FAIL" | "STYLE" | "REF" | "OTHER";

function classify(attrs: string, type: string, id: string): Category {
  const a = attrs.toLowerCase();

  // 1. Missing Alt Text (From Lua)
  if (type === "Image (No Alt)") return "ALT_FAIL";

  // 2. Highlights & Broken Text (Highest Priority)
  if (a.includes(".mark") || type === "Text (Raw)") return "MARK";
  
  // 3. Comments (Strict check to avoid matching "background" in text)
  if (a.includes("comment") || a.includes("annotation") || a.includes("background-color")) return "MARK";

  // 4. Styles (Underline, width, color, etc.)
  if (a.includes(".underline") || a.includes("style=") || a.includes("width=") || a.includes("color")) return "STYLE";

  // 5. References (Figures and Tables)
  if (id.startsWith("tbl-") || id.startsWith("fig-")) return "REF";

  return "OTHER";
}

// --- MAIN EXECUTION ---
console.log(`Auditing Attributes in: ${reportsDir}`);
console.log("Legend:\n 🚨 \x1b[31mHighlights/Broken/No-Alt\x1b[0m\n 🟠 \x1b[33mStyling\x1b[0m\n 📘 \x1b[34mReferences\x1b[0m\n ⚪ Structure\n");

// Ensure Lua filter exists
try { await Deno.stat(luaFilter); } catch { console.error(`Error: Lua filter not found at ${luaFilter}`); Deno.exit(1); }

const fileResults = new Map<string, Record<Category, string[]>>();

for await (const entry of walk(reportsDir)) {
  if (entry.isFile && entry.name.endsWith(".qmd")) {
    
    const cmd = new Deno.Command("quarto", {
      args: ["pandoc", entry.path, "--from", "markdown", "--lua-filter", luaFilter, "--to", "native"],
      stdout: "null", stderr: "piped",
    });

    const { stderr } = await cmd.output();
    const rawOutput = new TextDecoder().decode(stderr).trim();

    if (rawOutput) {
      const relPath = relative(join(scriptDir, ".."), entry.path);
      
      const buckets: Record<Category, string[]> = { MARK: [], ALT_FAIL: [], STYLE: [], REF: [], OTHER: [] };
      let hasContent = false;

      rawOutput.split("\n").forEach(line => {
        const parts = line.split("|");
        if (parts.length >= 4) {
          const [_, type, attrs, content] = parts;
          
          // Extract ID from attrs string "{#my-id .class}" -> "my-id"
          const idMatch = attrs.match(/#([\w-]+)/);
          const id = idMatch ? idMatch[1] : "";

          const cat = classify(attrs, type, id);
          
          let icon = "⚪";
          if (cat === "MARK" || cat === "ALT_FAIL") icon = "\x1b[31m🚨\x1b[0m"; // Red
          if (cat === "STYLE") icon = "\x1b[33m🟠\x1b[0m"; // Yellow
          if (cat === "REF")   icon = "\x1b[34m📘\x1b[0m"; // Blue
          
          const display = `   ${icon} [${type}] ${attrs} \x1b[2m"${content}"\x1b[0m`;
          buckets[cat].push(display);
          hasContent = true;
        }
      });

      if (hasContent) fileResults.set(relPath, buckets);
    }
  }
}

// --- DISPLAY RESULTS ---
if (fileResults.size === 0) console.log("No attributes found.");
else {
  const sortedFiles = Array.from(fileResults.keys()).sort();
  for (const file of sortedFiles) {
    const buckets = fileResults.get(file)!;
    
    // Check if we have anything worth showing
    const total = Object.values(buckets).reduce((acc, arr) => acc + arr.length, 0);
    if (total > 0) {
      console.log(`\x1b[1m📄 ${file}\x1b[0m`);
      
      // ORDER MATTERS HERE:
      printBucket("Highlights, Comments & Missing Alt", [...buckets.ALT_FAIL, ...buckets.MARK]);
      printBucket("Styling (Underlines, Widths)", buckets.STYLE);
      printBucket("Figures & Tables", buckets.REF);
      printBucket("Structure & Other", buckets.OTHER);
      console.log(""); 
    }
  }
}

function printBucket(title: string, items: string[]) {
  if (items.length > 0) {
    console.log(`   \x1b[4m${title}\x1b[0m`);
    items.forEach(i => console.log(i));
  }
}

File: _scripts/audit-div-span-attributes.lua
-- _scripts/audit-div-span-attributes.lua

-- Helper to safely extract attributes
local function get_attr_str(elem)
  if not elem.attr then return nil end
  
  local parts = {}
  if elem.attr.identifier ~= "" then table.insert(parts, "#" .. elem.attr.identifier) end
  for _, c in ipairs(elem.attr.classes) do table.insert(parts, "." .. c) end
  for k, v in pairs(elem.attr.attributes) do table.insert(parts, k .. "=\"" .. v .. "\"") end

  if #parts > 0 then return "{" .. table.concat(parts, " ") .. "}" end
  return nil
end

local function emit(file, type, attr, content)
  content = content:gsub("\n", " ")
  if #content > 60 then content = content:sub(1, 57) .. "..." end
  io.stderr:write(string.format("%s|%s|%s|%s\n", file, type, attr, content))
end

-- 1. Check Standard Elements
local function check_element(type, elem)
  local attr_str = get_attr_str(elem)
  local file = PANDOC_STATE.input_files[1] or "Unknown"
  local content = ""

  -- Special handling for Images (Check Alt Text)
  if type == "Image" then
    content = pandoc.utils.stringify(elem.caption)
    if content == "" then
      -- Emit specific warning for missing alt text, even if it has no attributes
      emit(file, "Image (No Alt)", attr_str or "{}", "MISSING ALT TEXT")
      return -- Stop here to avoid double printing
    end
  end

  if attr_str then
    if not content or content == "" then
      if elem.content then content = pandoc.utils.stringify(elem.content)
      elseif elem.text then content = elem.text
      end
    end
    emit(file, type, attr_str, content)
  end
end

-- 2. Check Plain Text for "Broken" Attributes
local function check_text(elem)
  local s = elem.text
  if s:match("{[%w%-%._#=]+}") then
    local file = PANDOC_STATE.input_files[1] or "Unknown"
    emit(file, "Text (Raw)", "Potential Broken Attribute", s)
  end
end

return {
  Span      = function(e) check_element("Span", e) end,
  Div       = function(e) check_element("Div", e) end,
  Link      = function(e) check_element("Link", e) end,
  Image     = function(e) check_element("Image", e) end,
  Code      = function(e) check_element("Code", e) end,
  CodeBlock = function(e) check_element("CodeBlock", e) end,
  Header    = function(e) check_element("Header", e) end,
  Str       = function(e) check_text(e) end
}
